### ISSUE TITLE
Nested macros stop working when ecto updated from 2.1.5 to 2.2.3

### ISSUE THREAD

**Issue by bobek** [elixir-ecto/ecto#2217]
I am filling this as a bug report as bugreport as problem appears after updating Ecto. I am far from being Elixir/Ecto expert, so please excuse my ignorance.

### Environment

* Elixir version (elixir -v): `Elixir 1.5.1`
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): `PostgreSQL 9.6.5`
* Ecto version (mix deps): `2.2.3`
* Database adapter and version (mix deps): `postgrex 0.13.3`
* Operating system: `Debian GNU/Linux sid`

### Current behavior

I am playing with Tilex, and https://github.com/hashrocket/tilex/blob/master/lib/tilex/stats.ex is failing to compile if Ecto is updated to `2.2.3`. It will through the following error:

```
== Compilation error in file lib/tilex/stats.ex ==
** (ArgumentError) argument error
    :erlang.apply(%Macro.Env{aliases: [{SQL, Ecto.Adapters.SQL}, {Repo, Tilex.Repo}], context: nil, context_modules: [Tilex.Stats], export_vars: nil, file: "/home/bobek/temp/elixir/tilex/lib/tilex/stats.ex", function: {:hot_posts, 0}, functions: [{Ecto.Query, [exclude: 2, first: 1, first: 2, last: 1, last: 2, subquery: 1, subquery: 2]}, {Kernel, [!=: 2, !==: 2, *: 2, +: 1, +: 2, ++: 2, -: 1, -: 2, --: 2, /: 2, <: 2, <=: 2, ==: 2, ===: 2, =~: 2, >: 2, >=: 2, abs: 1, apply: 2, apply: 3, binary_part: 3, bit_size: 1, byte_size: 1, div: 2, elem: 2, exit: 1, function_exported?: 3, get_and_update_in: 3, get_in: 2, hd: 1, inspect: 1, inspect: 2, is_atom: 1, is_binary: 1, is_bitstring: 1, is_boolean: 1, is_float: 1, is_function: 1, is_function: 2, is_integer: 1, is_list: 1, ...]}], lexical_tracker: #PID<0.230.0>, line: 90, macro_aliases: [], macros: [{Ecto.Query, [distinct: 2, distinct: 3, dynamic: 1, dynamic: 2, from: 1, from: 2, group_by: 2, group_by: 3, having: 2, having: 3, join: 3, join: 4, join: 5, limit: 2, limit: 3, lock: 2, offset: 2, offset: 3, or_having: 2, or_having: 3, or_where: 2, or_where: 3, order_by: 2, order_by: 3, preload: 2, preload: 3, select: 2, select: 3, select_merge: 2, select_merge: 3, update: 2, update: 3, where: 2, where: 3]}, {Kernel, [!: 1, &&: 2, ..: 2, <>: 2, @: 1, alias!: 1, and: 2, binding: 0, binding: 1, def: 1, def: 2, defdelegate: 2, defexception: 1, defimpl: 2, defimpl: 3, defmacro: 1, defmacro: 2, defmacrop: 1, defmacrop: 2, defmodule: 2, defoverridable: 1, defp: 1, defp: 2, defprotocol: 2, defstruct: 1, destructure: 2, get_and_update_in: 2, if: 2, in: 2, is_nil: 1, match?: 2, or: 2, pop_in: 1, put_in: 2, raise: 1, raise: 2, reraise: 2, ...]}], match_vars: :warn, module: Tilex.Stats, prematch_vars: nil, requires: [Ecto.Query, Kernel, Kernel.Typespec], vars: []}, :module, [])
    (elixir) lib/macro.ex:1196: Macro.do_expand_once/2
    (elixir) lib/macro.ex:1252: Macro.expand_until/2
    lib/ecto/query/builder.ex:490: Ecto.Query.Builder.try_expansion/5
    (elixir) lib/enum.ex:1357: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    lib/ecto/query/builder.ex:81: Ecto.Query.Builder.escape/5
    lib/ecto/query/builder/select.ex:126: anonymous fn/4 in Ecto.Query.Builder.Select.escape_pairs/4
    (elixir) lib/enum.ex:1357: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/enum.ex:1357: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    lib/ecto/query/builder/select.ex:69: Ecto.Query.Builder.Select.escape/4
    lib/ecto/query/builder/select.ex:204: Ecto.Query.Builder.Select.build/5
    expanding macro: Ecto.Query.select/3
    lib/tilex/stats.ex:90: Tilex.Stats.hot_posts/0
    expanding macro: Ecto.Query.from/2
    lib/tilex/stats.ex:90: Tilex.Stats.hot_posts/0
    (elixir) lib/kernel/parallel_compiler.ex:121: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/1
```
It compiles with Ecto version `2.1.5`. Offending line is
```
         hours_age: greatest(hours_since(p.published_at), 0.1)
```
which uses 2 macros (`greatest` and `hours_since`). 
```elixir
  defmacro greatest(value1, value2) do
    quote do
      fragment("greatest(?, ?)", unquote(value1), unquote(value2))
    end
  end

  defmacro hours_since(timestamp) do
    quote do
      fragment(
        "extract(epoch from (current_timestamp - ?)) / 3600",
        unquote(timestamp)
      )
    end
  end
```
When rewritten to
```
         hours_age: fragment("greatest(?,?)", hours_since(p.published_at), 0.1)
```
compilation is successful (and query to DB looks as expected).

### Expected behavior

Nesting of macros would work as before.

**Comment by josevalim** [comment]
Thank you! Fixed on v2.2 and master branches. A new release should be out soon, just waiting for approval from @michalmuskala.

**Comment by OvermindDL1** [comment]
Whoops, looks like I might have introduced that when I fixed the macro's fully expanding when they shouldn't have.  There was no test-case for this issue in the tests so I never saw it.  ^.^;

Still no test case it seems?
